packages/dds/tree/src/simple-tree/tree.ts

```diff
@@ -162,39 +283,57 @@ export function walkNodeSchema(
 		// nothing to do
 	} else if (isObjectNodeSchema(schema)) {
 		for (const field of schema.fields.values()) {
-			walkAllowedTypes(field.allowedTypeSet, visitor, visitedSet);
+			walkFieldSchema(field, visitor, visitedSet);
 		}
 	} else {
 		assert(
 			schema.kind === NodeKind.Array || schema.kind === NodeKind.Map,
 			0x9b3 /* invalid schema */,
 		);
 		const childTypes = schema.info as ImplicitAllowedTypes;
-		walkFieldSchema(childTypes, visitor, visitedSet);
+		walkAllowedTypes(normalizeAllowedTypes(childTypes), visitor, visitedSet);
 	}
 	// This visit is done at the end so the traversal order is most inner types first.
 	// This was picked since when fixing errors,
 	// working from the inner types out to the types that use them will probably go better than the reverse.
 	// This does not however ensure all types referenced by a type are visited before it, since in recursive cases thats impossible.
-	visitor(schema);
+	visitor.node?.(schema);
 }
 
 export function walkFieldSchema(
 	schema: ImplicitFieldSchema,
-	visitor: (schema: TreeNodeSchema) => void,
+	visitor: SchemaVisitor,
 	visitedSet: Set<TreeNodeSchema> = new Set(),
 ): void {
 	walkAllowedTypes(normalizeFieldSchema(schema).allowedTypeSet, visitor, visitedSet);
 }
 
 export function walkAllowedTypes(
 	allowedTypes: Iterable<TreeNodeSchema>,
-	visitor: (schema: TreeNodeSchema) => void,
+	visitor: SchemaVisitor,
 	visitedSet: Set<TreeNodeSchema>,
 ): void {
 	for (const childType of allowedTypes) {
 		walkNodeSchema(childType, visitor, visitedSet);
 	}
+	visitor.allowedTypes?.(allowedTypes);
+}
+
+/**
+ * Callbacks for use in {@link walkFieldSchema} / {@link walkAllowedTypes} / {@link walkNodeSchema}.
+ */
+export interface SchemaVisitor {
+	/**
+	 * Called once for each node schema.
+	 */
+	node?: (schema: TreeNodeSchema) => void;
```

Comment:
Presumably `undefined` for both of these is equivalent to a no-op? Very much a nitpick, but I like calling that distinction out explicitly in the docs, so users don't have to wonder if there is a special meaning to `undefined`.